src/client/runner/automation/util.js

```diff
@@ -11,27 +13,102 @@ var positionUtils   = testCafeCore.positionUtils;
+    if (isTextEditable || isContentEditable) {
+        if (isContentEditable && isNaN(parseInt(caretPos, 10)))
+            textSelection.setCursorToLastVisiblePosition(element);
+        else {
+            var position = isNaN(parseInt(caretPos, 10)) ? element.value.length : caretPos;
+
+            textSelection.select(element, position, position);
+        }
+    }
+    else {
+        // NOTE: if focus is called for not contentEditable element (like 'img' or 'button') inside
+        // contentEditable parent we should try to set right window selection. Generally we can't
+        // set right window selection object because after selection setup window.getSelection method
+        // returns  a different object depending on the browser.
+        var contentEditableParent = contentEditable.findContentEditableParent(element);
+
+        if (contentEditableParent)
+            textSelection.setCursorToLastVisiblePosition(contentEditable.findContentEditableParent(contentEditableParent));
+    }
+}
+
+export function focusAndSetSelection (element, simulateFocus, caretPos) {
+    return new Promise(resolve => {
+        var activeElement               = domUtils.getActiveElement();
+        var isTextEditable              = domUtils.isTextEditableElement(element);
+        var labelWithForAttr            = domUtils.closest(element, 'label[for]');
+        var shouldFocusByRelatedElement = !domUtils.isElementFocusable(element) && labelWithForAttr;
+        var isContentEditable           = domUtils.isContentEditableElement(element);
+        var elementForFocus             = isContentEditable ? contentEditable.findContentEditableParent(element) : element;
+
+        // NOTE: in WebKit if selection was never set in an input element focus method selects all text
+        // of this element. So we should select call select before focus to set caret to first symbol.
+        if (simulateFocus && browserUtils.isWebKit && isTextEditable)
+            textSelection.select(element, 0, 0);
+
+        // NOTE: we should call focus for element related with 'label' with attribute 'for'
+        if (shouldFocusByRelatedElement) {
+            if (simulateFocus)
+                focusByRelatedElement(labelWithForAttr);
+
+            resolve();
+            return;
+        }
+
+        var focusWithSilentMode = !simulateFocus;
+        var focusForMouseEvent  = true;
+
+        focusBlurSandbox.focus(elementForFocus, () => {
+            // NOTE: if another element was focused in the focus event handler we should not set selection
+            if (simulateFocus && !isContentEditable && element !== domUtils.getActiveElement()) {
+                resolve();
+                return;
+            }
+
+            setCaretPosition(element, caretPos);
+
+            // NOTE: we can't avoid element focusing because set selection methods lead to focusing.
+            // So we just focus previous active element without handlers if we don't need focus here
```

Comment:
```
            // NOTE: we can't avoid the element being focused because the setSelection method leads to focusing.
            // So, we just focus the previous active element without handlers if we don't need focus here
```